File: Data Structures & Algorithms/combination-target-sum/submission-5.cpp
class Solution {
public:
    void dfs(vector<int>& nums, vector<vector<int>>& res, int target, vector<int> cur, int index){
        if(target == 0){
            res.push_back(cur);
            return;
        }
        
        for(int j = index; j < nums.size(); j++){
            if(target - nums[j] < 0)
                continue;

            cur.push_back(nums[j]);
            dfs(nums, res, target - nums[j], cur, j);

            cur.pop_back();    
        }
    }

    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<vector<int>> res;
        dfs(nums, res, target, {}, 0);
        return res;
    }
};

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-5.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l = 0
        r = len(nums) - 1
        while l < r:
            mid = l + (r - l) // 2
            if nums[mid] > nums[r]:
                l = mid + 1
            else:
                r = mid
        pivot = l
        if target == nums[pivot]:
            return pivot
        if target == nums[0]:
            return 0
        if target == nums[len(nums)-1]:
            return len(nums) - 1
        l = 0; r = -1
        
        if target > nums[0] and target < nums[pivot]:
            l = 0
            r = pivot
        elif target > nums[pivot] and target < nums[len(nums)-1]:
            l = pivot
            r = len(nums) - 1
        while l <= r:
            mid = l + (r - l) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] < target:
                l = mid + 1
            else:
                r = mid - 1
        return -1